test/lib/hierarchy-operations/lineage-constraints.spec.js

```diff
@@ -1,82 +1,91 @@
+    it('no parent type yields undefined error', async () => expect(runScenario([], 'person')).to.eventually.rejectedWith('undefined'));
 
-    it('no valid parents yields not defined', async () => expect(await scenario([{
+    it('no valid parents yields not defined', async () => expect(runScenario([{
       id: 'person',
       parents: ['district_hospital'],
-    }], 'person', 'health_center')).to.include('cannot have parent of type'));
+    }], 'person', 'health_center')).to.eventually.rejectedWith('cannot have parent of type'));
 
     it('no settings doc requires valid parent type', async () => {
-      const mockDb = { get: () => { throw { name: 'not_found' }; } };
-      const { getHierarchyErrors } = await lineageConstraints(mockDb, { type: 'dne' });
-      const actual = getHierarchyErrors({ type: 'person' });
-      expect(actual).to.include('cannot have parent of type');
+      const mockDb = { get: () => { throw { status: 404 }; } };
+      const { assertNoHierarchyErrors } = await lineageConstraints(mockDb, { merge: false });
+      const actual = () => assertNoHierarchyErrors([{ type: 'person' }], { type: 'dne' });
+      expect(actual).to.throw('cannot have parent of type');
     });
 
     it('no settings doc requires valid contact type', async () => {
-      const mockDb = { get: () => { throw { name: 'not_found' }; } };
-      const { getHierarchyErrors } = await lineageConstraints(mockDb, { type: 'clinic' });
-      const actual = getHierarchyErrors({ type: 'dne' });
-      expect(actual).to.include('unknown type');
+      const mockDb = { get: () => { throw { status: 404 }; } };
+      const { assertNoHierarchyErrors } = await lineageConstraints(mockDb, { merge: false });
+      const actual = () => assertNoHierarchyErrors({ type: 'dne' }, { type: 'clinic' });
+      expect(actual).to.throw('unknown type');
     });
 
     it('no settings doc yields not defined', async () => {
-      const mockDb = { get: () => { throw { name: 'not_found' }; } };
-      const { getHierarchyErrors } = await lineageConstraints(mockDb, { type: 'clinic' });
-      const actual = getHierarchyErrors({ type: 'person' });
+      const mockDb = { get: () => { throw { status: 404 }; } };
+      const { assertNoHierarchyErrors } = await lineageConstraints(mockDb, { merge: false });
+      const actual = assertNoHierarchyErrors({ type: 'person' }, { type: 'clinic' });
       expect(actual).to.be.undefined;
     });
 
+    it('cannot merge with self', async () => {
+      await expect(runScenario([], 'a', 'a', true)).to.eventually.rejectedWith('self');
+    });
+
+    it('cannot merge with id: "root"', async () => {
+      const mockDb = { get: () => ({ settings: { contact_types: [] } }) };
+      const { assertNoHierarchyErrors } = await lineageConstraints(mockDb, { merge: true });
+      const actual = () => assertNoHierarchyErrors({ _id: 'root', type: 'dne' }, { _id: 'foo', type: 'clinic' });
+      expect(actual).to.throw('root');
+    });
+
```

Comment:
suggestion: one additional merge case:

```suggestion

    it('cannot merge different types', async () => {
      const sourceType = 'person';
      const destinationType = 'health_center';
      const actual = runScenario([{
        id: 'person',
        parents: ['health_center'],
      }], sourceType, destinationType, true);

      await expect(actual).to.eventually.rejectedWith(
        `Hierarchy Constraints: source and destinations must have the same type. Source is "${sourceType}" while destination is "${destinationType}".`
      );
    });

```